### DISCUSSION TITLE
Question about the update in Lagrangian PID update

### DISCUSSION THREAD

**Discussion by bigscientist** [PKU-Alignment/omnisafe#335] [General]
In PIDLagrangian.py, the script to get "D" is pid_d = max(0.0, self._cost_d - self._cost_ds[0]), 
I wonder that "D" is the difference between neighbour value of Lagrangian and it shall be 
pid_d = max(0.0, self._cost_d - self._cost_ds[-1])
![explorer_dSumrNJrT8](https://github.com/PKU-Alignment/omnisafe/assets/48021319/7e2664d7-3d9f-43d4-8755-1a78551d8243)

**Comment by Gaiejj** [comment]
Very careful observation! In fact, this code is not written incorrectly. It was initially designed to use the first item in the _cost_ds queue to solve the `pid_d`. In our experiments, we found that adding this delay mechanism makes the algorithm update more smoothly. You can cancel this delay by setting the `pid_d_delay` parameter to 1 in `omnisafe/configs/on-policy/CPPOPID.yaml`.

**Reply by bigscientist** [reply]
Thank you, I get it
